kie-server-parent/kie-server-services/kie-server-services-openshift/src/main/java/org/kie/server/services/impl/OpenShiftStartupStrategy.java

```diff
@@ -0,0 +1,209 @@
+/*
+ * Copyright 2018 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+*/
+
+package org.kie.server.services.impl;
+
+import java.time.ZonedDateTime;
+import java.time.format.DateTimeFormatter;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.UUID;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Supplier;
+
+import io.fabric8.kubernetes.api.model.ConfigMap;
+import io.fabric8.kubernetes.api.model.ConfigMapBuilder;
+import io.fabric8.kubernetes.api.model.ObjectMeta;
+import io.fabric8.kubernetes.client.KubernetesClientException;
+import io.fabric8.kubernetes.client.Watch;
+import io.fabric8.kubernetes.client.Watcher;
+import io.fabric8.openshift.api.model.DeploymentConfig;
+import io.fabric8.openshift.client.OpenShiftClient;
+import org.kie.server.api.KieServerConstants;
+import org.kie.server.api.model.KieContainerResource;
+import org.kie.server.controller.api.model.KieServerSetup;
+import org.kie.server.services.api.StartupStrategy;
+import org.kie.server.services.impl.storage.KieServerState;
+import org.kie.server.services.impl.storage.cloud.CloudClientFactory;
+import org.kie.server.services.impl.storage.cloud.KieServerStateCloudRepository;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class OpenShiftStartupStrategy implements StartupStrategy {
+
+    private static final String KIE_SERVER_ROLLOUT_IN_PROGRESS = "kieserver-rollout-in-progress";
+    private static final String KIE_SERVER_INSTANCE_ID = "kieserver-" + UUID.randomUUID().toString();
+    private static final String ROLLOUT_TRIGGER_TIMESTAMP = "services.server.kie.org/openshift-startup-strategy.changeTimestamp";
+    private static final Logger logger = LoggerFactory.getLogger(OpenShiftStartupStrategy.class);
+
+    private static Supplier<OpenShiftClient> clouldClientHelper = () -> 
+        (OpenShiftClient) (new CloudClientFactory() {}).createAdaptiveClient();
```

Comment:
in context of the other comment would be more safe to have the client to be already known to be OpenShiftClient so we could avoid the cast...

Reply:
Refactored CloudClientFactory and eliminated the needs for casting.